org.esbtools.auth.common/src/main/java/org/esbtools/auth/servlet/CertEnvironmentVerificationFilter.java

```diff
@@ -0,0 +1,75 @@
+package org.esbtools.auth.servlet;
+
+import java.io.IOException;
+import java.security.cert.X509Certificate;
+
+import javax.naming.NamingException;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+
+import org.esbtools.auth.util.Environment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CertEnvironmentVerificationFilter implements Filter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CertEnvironmentVerificationFilter.class);
+
+  private final Environment env;
+
+  public CertEnvironmentVerificationFilter(String environment) {
+    env = new Environment(environment);
+
+    LOGGER.info("Cert Environment: " + ((environment == null) ? "Not Set" : environment));
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+    //Do Nothing!
+  }
+
+  @Override
+  public void destroy() {
+    //Do Nothing!
+  }
+  
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+    LOGGER.debug("Attempting Environment Cert verification");
+    X509Certificate certChain[] = (X509Certificate[]) request.getAttribute("javax.servlet.request.X509Certificate");
+
+    if ((null != certChain) && (certChain.length > 0)) {
```

Comment:
Doesn't this line ensure that the environment validation check will never happen if the client doesn't present a certificate?

Reply:
You are correct, that is exactly what it does. The filter itself fires every time, regardless of if a cert is provided or not. But the validation only happens if a cert is present.

I guess I think of CertEnvironmentVerificationFilter as being akin to CertLdapLoginModule. But perhaps it is more nuanced then that?